### PR TITLE
test-threads refactor

### PR DESCRIPTION
**Pull request by barrettk**
- now relies on `nmrec` instead of string manipulation
 - we should require a miniumum version of nmrec once it's officially released

addresses one of the bullets in https://github.com/metrumresearchgroup/bbr/issues/599

### CODE REVIEW

**Comment by barrettk** [comment]
@kyleam This PR is ready to review with the exception of being able to fetch `nmrec` and require a minimum version in the DESCRIPTION file

**Comment by kyleam** [comment]
Okay, thanks.  I'm working on $PROB support for nmrec.  That shouldn't take long, so I'd like to get that into the next nmrec release.

Two initial thoughts :

 * Using nmrec in this case can essentially be viewed as a fix.  It's a good thing if the current tests pass for you, but shouldn't this also add regression cases for at least a few examples that were broken with the old code?

 * Before this is merged, we should discuss (offline) how to handle nmrec as a dependency: <https://github.com/metrumresearchgroup/bbr/issues/599#issuecomment-1613813457>.  It's probably best to mark this as a draft until then, even when it's ready for a full review.  (What we decide in that discussion main mean that more than an nmrec release would block the merge here.)

Also, if you'd like to get a passing drone before the final adjustments, you can temporarily add something like

```
  - git clone --depth 1 https://github.com/metrumresearchgroup/nmrec.git /tmp/nmrec
  - R -s -e 'devtools::install("/tmp/nmrec", upgrade = "never")'
```

to the drone install steps.

**Comment by barrettk** [comment]
> * Using nmrec in this case can essentially be viewed as a fix.  It's a good thing if the current tests pass for you, but shouldn't this also add regression cases for at least a few examples that were broken with the old code?

I initially didn't think there were any cases where the old code would have broken, but after thinking on it a bit there definitely were - mainly models that used one of the weird spellings of `maxevals` or something. We captured all the frequent cases with regex, but definitely did not support all of them (dont think we supported lower case either). Can add a test for this type of thing unless there were any other cases you had in mind. Im changing some of the tests now to rely on nmrec as well, so our existing tests will be a bit better.


> * Before this is merged, we should discuss (offline) how to handle nmrec as a dependency: [Refactor current functions to use `nmrec` #599 (comment)](https://github.com/metrumresearchgroup/bbr/issues/599#issuecomment-1613813457).  It's probably best to mark this as a draft until then, even when it's ready for a full review.  (What we decide in that discussion main mean that more than an nmrec release would block the merge here.)

Yeah this sounds good. As an aside, my intention was just to address the test-threads changes in this PR, but we could broaden the scope to include all the changes mentioned in that issue as well.


>Also, if you'd like to get a passing drone before the final adjustments, you can temporarily add something like
```
  - git clone --depth 1 https://github.com/metrumresearchgroup/nmrec.git /tmp/nmrec
  - R -s -e 'devtools::install("/tmp/nmrec", upgrade = "never")'
```

Thanks, will try this out once I fix the tests

**Comment by barrettk** [comment]
@kyleam FYI it seems like it's failing solely because of `gregexec` in `nmrec:::split_to_elements`, which is not included in R 4.0. I have reported a [similar issue](https://github.com/davidgohel/flextable/issues/457) in the past for an MPN package, so im wondering if we want to make a fix for this on the `nmrec` side

**Comment by kyleam** [comment]
> FYI it seems like it's failing solely because of `gregexec` in `nmrec:::split_to_elements`

Thanks, didn't realize that was a recent addition.  Fixed in https://github.com/metrumresearchgroup/nmrec/pull/11

**Comment by kyleam** [comment]
> We captured all the frequent cases with regex, but definitely did not support all of them (dont think we supported lower case either).

Hmm, I'm not as confident as you are about what's frequent, but, yeah, the idea is that, by using nmrec, you support all the spellings, as well as little things that the previous bbr code doesn't consider.  For example, the option name could be `NBURN = 1000` or even `NBURN 1000`.

> Can add a test for this type of thing unless there were any other cases you had in mind.

Thanks.  I don't care about any particular case (and I don't think it's necessary to exhaustively cover weird cases; that's nmrec's job) but having a few cases that wouldn't work before this change is a good thing in my view.

> As an aside, my intention was just to address the test-threads changes in this PR, but we could broaden the scope to include all the changes mentioned in that issue as well.

My vote is to keep this PR focused on test-threads.

**Comment by barrettk** [comment]
@kyleam got drone to work by re-running it. Seems like we will wait to actually merge this until nmrec is officially released, but figured i'd still ask for a review now (no rush).

File: R/test-threads.R
Comment (kyleam) [inline R/test-threads.R:347]: This and the similar lines below are fine as is, but I'm guessing they came before the switch to making `get_record_option` return `NULL` in nmrec's [47228bd](https://github.com/metrumresearchgroup/nmrec/commit/47228bddb5736f371781c25bd48ac63daa6ca476). With that, this could also be written as is

```R
max_detect <- !is.null(get_record_option(est_block.i, "maxevals"))
```

But, again, I think either way is okay.

File: R/test-threads.R
Comment (kyleam) [inline R/test-threads.R:361]: All of the other spots that call `nmrec::set_record_option` are modifying an existing option, so they keep whatever name the caller used originally, which is good. This spot is adding a new option, so the spelling you give will go in.

Makes sense, and I like that you took care to go with upper case here because it's my impression that that's the predominant style for NONMEM options.

File: tests/testthat/test-test-threads.R
Comment (kyleam) [inline tests/testthat/test-test-threads.R:29]: Okay as is, but `mod_lines` is only used in the `parse_ctl` call, so you could just use `read_ctl` directly.

File: tests/testthat/test-test-threads.R
Comment (kyleam) [inline tests/testthat/test-test-threads.R:46]: Leftover dead code?

File: tests/testthat/test-test-threads.R
Comment (kyleam) [inline tests/testthat/test-test-threads.R:140]: This is no longer confirming "the estimation method didn't change", right? (I think that's okay, but comment should be updated.)

Same comment for line 163.

File: tests/testthat/test-test-threads.R
Comment (kyleam) [inline tests/testthat/test-test-threads.R:191]: What's "recursive" referring to?

File: tests/testthat/test-test-threads.R
Comment (kyleam) [inline tests/testthat/test-test-threads.R:178]: Hmm, there aren't two "BAYES" methods in acop-fake-bayes.ctl:

<https://github.com/metrumresearchgroup/bbr/blob/feddd576f3f8cf9f091487b761391dedba481040/inst/model/nonmem/complex/acop-fake-bayes.ctl#L38-L40>

So I guess `get_est_options` *before* this PR was returning an incorrect result. Is that related to why the 0 and 10 swap places in the expected value?

File: tests/testthat/test-test-threads.R
Comment (kyleam) [inline tests/testthat/test-test-threads.R:190]: Nice, thanks for adding tests that would fail before this PR's changes.

File: tests/testthat/test-test-threads.R
Comment (kyleam) [inline tests/testthat/test-test-threads.R:223]: That `list` is unnecessary, right? At a glance, think that can just be `c(mod_complex_edit, mods_fake)`.

Comment (kyleam) [inline tests/testthat/test-test-threads.R:223]: Unlikely to actually bite us, but, for package code, better to use `TRUE` than `T` shorthand because `T` may be overridden to anything (set `T <- FALSE` for maximum confusion :x)

**Review by kyleam — APPROVED** [review]
The reworked approach looks good, and you fit it nicely into the previous existing conditions, making it easy to review regarding whether it preserves the existing intended logic.

The test code took me a bit longer to wrap my head around (and I'm not sure I completely succeeded), but I think that's mostly a matter of needing to refresh on what they were doing, not the propose changes (which in my view make things simpler and easier to understand). I left a few questions and comments on the tests, but nothing that I consider a blocker.

File: R/test-threads.R
Comment (barrettk) [reply R/test-threads.R:361]: Yeah I was conflicted with the case - thought about determining the case of other options, but that seemed a bit overkill

File: tests/testthat/test-test-threads.R
Comment (barrettk) [reply tests/testthat/test-test-threads.R:178]: No it was correct before, but was formatted differently and IMO more confusingly. It would attach the method to each of the options, so both of those expectations corresponded to the same "BAYES" method. The flipping of the 10 and 0 is just because the order of the options changed

File: tests/testthat/test-test-threads.R
Comment (barrettk) [reply tests/testthat/test-test-threads.R:223]: No `delete_models` expects model objects or lists:
```r
>     delete_models(c(mod_complex_edit, mods_fake), .force = T)
Error in check_model_object_list(.mods, .mod_types = c(NM_MOD_CLASS, NM_SUM_CLASS)) :
Passed list must contain only model objects, but found 5 invalid objects at indices: 1, 2, 3, 4, 5

>     delete_models(c(list(mod_complex_edit), mods_fake), .force = T)
Removed 3 models with the following tags:
- test threads
```

File: tests/testthat/test-test-threads.R
Comment (barrettk) [reply tests/testthat/test-test-threads.R]: Fixed

File: tests/testthat/test-test-threads.R
Comment (barrettk) [reply tests/testthat/test-test-threads.R]: Fixed

**Comment by kyleam** [comment]
> requested a review from [kyleam](https://github.com/kyleam) [now](https://github.com/metrumresearchgroup/bbr/pull/600#event-10298068177)

I've already approved.  Is there something in particular you want me to look at?  If not, merge if you're happy with its state.

File: tests/testthat/test-test-threads.R
Comment (kyleam) [reply tests/testthat/test-test-threads.R:178]: Hmm, I can't make heads or tails of it with a quick glance, but I'll take you word for it.  Thanks.

File: tests/testthat/test-test-threads.R
Comment (kyleam) [reply tests/testthat/test-test-threads.R]: Doh, because model object is a list underneath.  Thanks for the correction.

**Comment by barrettk** [comment]
> I've already approved. Is there something in particular you want me to look at? If not, merge if you're happy with its state.

Yeah fair - I made some changes based on some of your comments, and just wanted to make sure you were good with the changes. Will merge now

File: tests/testthat/test-test-threads.R
Comment (barrettk) [reply tests/testthat/test-test-threads.R]: Oh jeez yeah brain fart

**Pull request merged**
